### PR TITLE
Add RSS allocated service external IPs during Rack Initialization.

### PR DESCRIPTION
**Pull request by luqmana**
Aside from Nexus, we weren't adding the external IPs used by other services to the DB during rack initialization. Currently that's just External DNS and Boundary NTP. As part of that, instead of duplicating the separate `nexus_service` table for each service, updated `external_ip` to keep a primary key to either the instance or service table directly.

### CODE REVIEW

**Comment by luqmana** [comment]
> Sorry in advance for a rambling question but I don't yet fully grok the collapsing of InternalDns{,Config} into just InternalDns and ExternalDns{,Config} into just ExternalDns. I think my confusion is that I thought the "service" table in the database represented logically distinct "network services" in the control plane. (I'm using the term the way RFD 248 talks about them. Each service would correspond to a different DNS name where we'd host SRV records.) In that sense, the HTTP (configuration) service and the DNS service are two different things provided by the DNS zones. They'll have different DNS names resolving to different IP/port pairs. We discussed this a bit [here](https://github.com/oxidecomputer/omicron/pull/2753#issuecomment-1495142734). I think the real motivation at the time was that this was the only place we stored the IP and port information for these services and I thought this might become the canonical source of the DNS information. That's not true any more. Whatever provisions things also needs to update the `dns_names` table with the real details. So maybe this doesn't matter at all! But then I'm not clear on what the "service" table really represents. Is it "deployed components"? Should zone_id now be unique again?

Yea, the story there has been confusing and there were a couple points that led me to collapse them:

1. `*Dns` & `*DnsConfig` run in the same smf service and process which seemed at odds with the RFD 248 definitions where "instances" of a "service" are actually separate processes. Compare also Nexus which similarly has multiple (internal & external) endpoints. But the `address` and `port` we record in its `service` record is just for the internal api.
2. We expect the `address` column in the `service` table to [be an IPv6 address](https://github.com/oxidecomputer/omicron/blob/f77cf314d065ea98d706b0d400ba62520bc70638/nexus/db-model/src/service.rs#L22) on the rack subnet. That's fine for `*DnsConfig` and `InternalDns` but not `ExternalDns`. Inside the `oxz_external_dns` zone, the DNS server is bound to an OPTE private address (e.g., `172.30.x.x`) and is externally accessible via an IP derived from the service IP pool. Note this applies to Nexus' external API as well but that's been side stepped because we've only recorded the internal api as noted above.

This is what I see after deploying locally today on main:
```
root@[fd00:1122:3344:101::6]:32221/omicron> SELECT * FROM service;
                   id                  |         time_created          |         time_modified         |               sled_id                |               zone_id                |          ip           | port  |        kind
---------------------------------------+-------------------------------+-------------------------------+--------------------------------------+--------------------------------------+-----------------------+-------+----------------------
  0e0bc790-2f82-494b-a408-35c719dc5027 | 2023-05-17 20:10:30.884862+00 | 2023-05-17 20:10:30.884862+00 | fec8daf2-83d0-4a3d-a119-81bbe02cdba0 | 0e0bc790-2f82-494b-a408-35c719dc5027 | fd00:1122:3344:101::3 |  5353 | external_dns_config
  6327bc98-fe0f-4fe9-a904-1c664b616433 | 2023-05-17 20:10:30.912112+00 | 2023-05-17 20:10:30.915773+00 | fec8daf2-83d0-4a3d-a119-81bbe02cdba0 | 6327bc98-fe0f-4fe9-a904-1c664b616433 | fd00:1122:3344:1::1   |  5353 | internal_dns
  98f43c17-d5a8-4c54-8996-bad0b1b9d399 | 2023-05-17 20:10:30.908391+00 | 2023-05-17 20:10:30.908391+00 | fec8daf2-83d0-4a3d-a119-81bbe02cdba0 | 98f43c17-d5a8-4c54-8996-bad0b1b9d399 | fd00:1122:3344:101::5 | 12223 | oximeter
  b563d880-495a-4003-950e-8a5aa1e98725 | 2023-05-17 20:10:30.923211+00 | 2023-05-17 20:10:30.923211+00 | fec8daf2-83d0-4a3d-a119-81bbe02cdba0 | b563d880-495a-4003-950e-8a5aa1e98725 | fd00:1122:3344:101::c |   123 | ntp
  c3ff015a-b7c8-49d3-ba0a-f209a6463a5d | 2023-05-17 20:10:30.888462+00 | 2023-05-17 20:10:30.888462+00 | fec8daf2-83d0-4a3d-a119-81bbe02cdba0 | c3ff015a-b7c8-49d3-ba0a-f209a6463a5d | fd00:1122:3344:101::4 | 12221 | nexus
  f3b3d43c-f0e4-4426-b7ce-c447a885bd9f | 2023-05-17 20:10:30.919531+00 | 2023-05-17 20:10:30.919531+00 | fec8daf2-83d0-4a3d-a119-81bbe02cdba0 | f3b3d43c-f0e4-4426-b7ce-c447a885bd9f | fd00:1122:3344:101::b | 17000 | crucible_pantry
(6 rows)
```

Note [we haven't populated](https://github.com/oxidecomputer/omicron/blob/f77cf314d065ea98d706b0d400ba62520bc70638/sled-agent/src/rack_setup/service.rs#L642-L651) a row for `external_dns` but as mentioned in (2), what would the address be?
```
$ pfexec svcprop -z oxz_external_dns -p config external_dns
config/http_address astring [fd00:1122:3344:101::3]:5353
config/dns_address astring 172.30.0.5:53
$ grep -A2 internal_services_ip_pool_ranges smf/sled-agent/non-gimlet/config-rss.toml
[[internal_services_ip_pool_ranges]]
first = "192.168.1.20"
last = "192.168.1.22"
```

`InternalDns*` doesn't have this problem as both only listen on internal addresses:
```
$ pfexec svcprop -z oxz_internal_dns -p config internal_dns
config/http_address astring [fd00:1122:3344:1::1]:5353
config/dns_address astring [fd00:1122:3344:1::1]:53
```
But note there's no entry for `internal_dns_config`? That's because Sled Agent already didn't have the config/non-config distinction and ended up reusing the same `service_id` for both:
https://github.com/oxidecomputer/omicron/blob/f77cf314d065ea98d706b0d400ba62520bc70638/sled-agent/src/rack_setup/service.rs#L652-L671

My first pass was actually trying to split things up on the Sled Agent side as well but then you have 2 "different" services both modifying the same smf config, trying to be started, etc

(I also just realized, `InternalDnsConfig` & `InternalDns` having the same `service_id` meant the latter ended up overwriting the former but not completey, the port is wrong. I'll push a fix for that.)

> I'm still a little confused about what the stuff _in_ the service table represents now. It doesn't appear used for much today so maybe it doesn't matter.

For my purposes here, I needed some kind of "service" record to anchor other related resources in the DB (e.g. `external_ip`, `network_interface`).

> Should zone_id now be unique again?
Yea, for the current set of things we seem to be adding with this change it would be one "service" per zone. But `ServiceKind::Dendrite` and `ServiceKind::Tfport` would be a case where both exist in the same zone. (As for why we're not actually adding those today...that's another question 😅)
 
> Unrelatedly: how does this work for configurations like development where we use the same "external" IP address for Nexus and external DNS? Is it not a problem that we've "allocated" the same IP twice?

Huh, I wasn't aware of such a case? When running locally I do have them on separate IPs (e.g. DNS at `192.168.1.20`, Nexus at `192.168.1.21`).

File: common/src/sql/dbinit.sql [hunk 1]
```diff
@@ -1529,8 +1512,11 @@ CREATE TABLE omicron.public.external_ip (
     /* FK to the `ip_pool_range` table. */
     ip_pool_range_id UUID NOT NULL,
 
-    /* FK to the `instance` table. See the constraints below. */
-    instance_id UUID,
+    /* True if this IP is associated with a service rather than an instance. */
+    is_service BOOL NOT NULL,
```
[the remark this answers is not in the document]
Comment (luqmana) [reply common/src/sql/dbinit.sql:1516]: Maybe? (Partly just used a flag as trying to come up with a new name for such an enum was hard given `kind` was already taken 😅 -- Maybe `is_service` -> `kind` (`Service` / `Guest`) and existing `kind` -> `ip_kind`?

File: nexus/src/lib.rs [hunk 2]
```diff
@@ -231,11 +231,9 @@ impl nexus_test_interface::NexusServer for Server {
         let internal_services_ip_pool_ranges = services
             .iter()
             .filter_map(|s| match s.kind {
-                ServiceKind::Nexus { external_address: IpAddr::V4(addr) } => {
-                    Some(IpRange::V4(Ipv4Range::new(addr, addr).unwrap()))
-                }
-                ServiceKind::Nexus { external_address: IpAddr::V6(addr) } => {
-                    Some(IpRange::V6(Ipv6Range::new(addr, addr).unwrap()))
+                ServiceKind::ExternalDns { external_address }
```
[the remark this answers is not in the document]
Comment (luqmana) [reply nexus/src/lib.rs:234]: Yup as long as they're the same type

File: common/src/sql/dbinit.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (davepacheco) [reply common/src/sql/dbinit.sql:1515]: Other ideas:
- `attached_to_type` (and call `parent_id` `attached_to_id`)
- `parent_type` + `parent_id`

We can also leave the boolean for now.

**Comment by davepacheco** [comment]
> > Unrelatedly: how does this work for configurations like development where we use the same "external" IP address for Nexus and external DNS? Is it not a problem that we've "allocated" the same IP twice?
> 
> Huh, I wasn't aware of such a case? When running locally I do have them on separate IPs (e.g. DNS at `192.168.1.20`, Nexus at `192.168.1.21`).

I thought that if you go through https://github.com/oxidecomputer/omicron/blob/main/docs/how-to-run-simulated.adoc, they'll both wind up on 127.0.0.1.  I thought that was also true in the test suite.

**Comment by davepacheco** [comment]
Yeah, RFD 248 uses the example of different Unix processes being different backends for a given service.  I didn't mean to imply there that a given process can't provide multiple services.  I really think that _is_ what's going on with the DNS servers -- that is, there are two different sets of SRV records that are used by different consumers to do different things.  They just happen (well, not coincidentally) to use overlapping processes.  Anyway, if you think of the "services" table as "the set of [RFD 248-definition] backends", then it makes sense that Nexus's _external_ address and the external DNS server's external address wouldn't be there, since those are not services provided within the control plane.  It's also fine that we expect the addresses here to be IPv6 addresses, since anything provided within the control plane would be on IPv6.

But this is not to say that the services table _should_ correspond exactly with RFD 248 backends.  At one time I thought that made sense.  Now that internal DNS has a separate `dns_name` table that's assumed to be kept in sync with this table, that's not necessary.

It really feels to me like this confusion stems from a lack of a consumer for this data.  The only existing consumer I found was the DNS propagation stuff using it to list the DNS servers that it's supposed to keep updated.  For that purpose, it doesn't matter if these are the list of network services or _zones_ providing those services.  Your point makes sense that for "a thing that an IP is attached to", you really want a table of zones, not service backends.  The other use case that I'm guessing this table was built for was the rebalancer, which will say something like "load [or availability] requires that I provision another internal DNS server", which is another case where you want this to be a list of zones, not network services.

I think this approach makes sense.  It _is_ a subtle but important change so I'm glad we're writing it down.  I wonder if we should rename the table `service_zone`?  (Feel free to defer.  I do think it'd be a lot clearer, especially given how overloaded "service" is.)

**Comment by luqmana** [comment]
> > > Unrelatedly: how does this work for configurations like development where we use the same "external" IP address for Nexus and external DNS? Is it not a problem that we've "allocated" the same IP twice?
> > 
> > 
> > Huh, I wasn't aware of such a case? When running locally I do have them on separate IPs (e.g. DNS at `192.168.1.20`, Nexus at `192.168.1.21`).
> 
> I thought that if you go through https://github.com/oxidecomputer/omicron/blob/main/docs/how-to-run-simulated.adoc, they'll both wind up on 127.0.0.1. I thought that was also true in the test suite.

Ah right, the simulated case. It does work because Nexus gets IPv4 localhost and ExternalDns gets IPv6:
```
omicron-dev: setting up all services ...
log file: /tmp/omicron-dev-omicron-dev.11604.0.log
note: configured to log to "/tmp/omicron-dev-omicron-dev.11604.0.log"
omicron-dev: services are running.
omicron-dev: nexus external API:    127.0.0.1:12220
omicron-dev: nexus internal API:    [::1]:12221
omicron-dev: cockroachdb pid:       12122
omicron-dev: cockroachdb URL:       postgresql://root@127.0.0.1:55105/omicron?sslmode=disable
omicron-dev: cockroachdb directory: /tmp/.tmpBbC8qH
omicron-dev: external DNS name:     oxide-dev.test
omicron-dev: external DNS HTTP:     http://[::1]:46936
omicron-dev: external DNS:          [::1]:61526
omicron-dev:   e.g. `dig @::1 -p 61526 test-suite-silo.sys.oxide-dev.test`
omicron-dev: silo name:             test-suite-silo
omicron-dev: privileged user name:  test-privileged
```
```
root@127.0.0.1:55105/omicron> select * from external_ip;;
                   id                  |     name     | description  |         time_created          |         time_modified         | time_deleted |              ip_pool_id              |           ip_pool_range_id           | is_service |              parent_id               |   kind   |    ip     | first_port | last_port
---------------------------------------+--------------+--------------+-------------------------------+-------------------------------+--------------+--------------------------------------+--------------------------------------+------------+--------------------------------------+----------+-----------+------------+------------
  b3b18860-4a23-4d82-8697-d5b7983bec00 | nexus        | nexus        | 2023-05-18 19:44:03.325652+00 | 2023-05-18 19:44:03.325652+00 | NULL         | 81e34373-a463-4caf-8f90-955863be3f67 | 696cd5e9-f9cf-4852-a584-abd73e26a886 |    true    | 98fb0ce4-f60f-4e75-bc19-2237555143c2 | floating | 127.0.0.1 |          0 |     65535
  ff43ae7a-d2dc-4b6c-97f8-f4c8e91d14ea | external-dns | external_dns | 2023-05-18 19:44:03.287446+00 | 2023-05-18 19:44:03.287446+00 | NULL         | 81e34373-a463-4caf-8f90-955863be3f67 | 9187b3e3-9878-424f-870b-f0f9fd122931 |    true    | 6aeb9dec-02b4-414b-bc5b-7d9d823e3a9e | floating | ::1       |          0 |     65535
(2 rows)
```

File: nexus/db-model/src/external_ip.rs [hunk 3]
```diff
@@ -59,11 +59,12 @@ pub struct ExternalIp {
...
     pub ip_pool_range_id: Uuid,
+    pub is_service: bool,
     // This is Some(_) for:
-    //  - all instance SNAT IPs
+    //  - all instance/service SNAT IPs
     //  - all ephemeral IPs
-    //  - a floating IP attached to an instance.
-    pub instance_id: Option<Uuid>,
+    //  - a floating IP attached to an instance or service.
+    pub parent_id: Option<Uuid>,
```
[the remark this answers is not in the document]
Comment (luqmana) [reply nexus/db-model/src/external_ip.rs:67]: I think a [composite data type](https://www.cockroachlabs.com/docs/stable/create-type.html#create-a-composite-data-type) on the SQL side and an enum + appropriate impls on the Rust side would work.

Created https://github.com/oxidecomputer/omicron/issues/3152 for that.

File: nexus/db-queries/src/db/datastore/rack.rs [hunk 4]
```diff
@@ -571,18 +582,20 @@ impl DataStore {
         opctx.authorize(authz::Action::Read, &authz::DNS_CONFIG).await?;
 
         use crate::db::schema::external_ip::dsl as extip_dsl;
-        use crate::db::schema::nexus_service::dsl as nexus_dsl;
+        use crate::db::schema::service::dsl as service_dsl;
         type TxnError = TransactionError<()>;
         self.pool_authorized(opctx)
             .await?
             .transaction_async(|conn| async move {
-                // This is the rare case where we want to allow table scans.
```
[the remark this answers is not in the document]
Comment (luqmana) [reply nexus/db-queries/src/db/datastore/rack.rs:579]: Gave this a quick try but given it's used cross-crate, just adding a `#[cfg(test)]` is not enough.

**Review by davepacheco — APPROVED** [review]

**Pull request merged**